ansible/action_plugins/agnosticd_odcr.py

```diff
@@ -0,0 +1,480 @@
+from __future__ import (absolute_import, division, print_function)
+import datetime
+import re
+import pprint
+from ansible.errors import AnsibleError
+from ansible.plugins.action import ActionBase
+from ansible.utils.display import Display
+
+import boto3
+import botocore.exceptions
+#pylint: disable=invalid-name
+__metaclass__ = type
+#pylint: enable=invalid-name
+
+DOCUMENTATION = """
+        action: agnosticd_odcr
+        author: Guillaume Core <gucore@redhat.com>
+        version_added: "2.9"
+        short_description: Create AWS on-demand Capacity Reservations
+        description:
+            - Create reservations and return Regions / AZs / Ids
+        options:
+          _terms:
+            description: reservations to create
+            required: True
+          ttl:
+            description: The TTL to set in the reservations
+            required: False
+            type: int
+            default: 1h
+"""
+
+EXAMPLES = """
+- name: Create on-demand capacity reservations and save result
+  agnosticd_odcr:
+    data: "{{ agnosticd_aws_capacity_reservation }}"
+    aws_access_key_id: "{{ aws_access_key_id }}"
+    aws_secret_access_key: "{{ aws_secret_access_key }}"
+    ttl: 1h
+  register: r_odcr
+"""
+
+RETURN = """
+_raw:
+   description: Reservations
+   type: list
+"""
+
+display = Display()
+pp = pprint.PrettyPrinter(indent=2)
+
+def parse_time(time_str):
+    """Parse duration (str) and return timedelta
+
+    Return False if the string doesn't match"""
+
+    regex = re.compile(
+        r'((?P<days>\d+?)d)?((?P<hours>\d+?)h)?((?P<minutes>\d+?)m)?((?P<seconds>\d+?)s)?'
```

Comment:
The regex could be improved with string anchors `^` and `$`. As written it will match to empty string on `1H` ` 3d` (leading space). It will also match empty string currently.

```
>>> regex.match('')
<re.Match object; span=(0, 0), match=''>
>>> regex.match('3D')
<re.Match object; span=(0, 0), match=''>
>>> regex.match(' 3d')
<re.Match object; span=(0, 0), match=''>
>>> regex.match('3')
<re.Match object; span=(0, 0), match=''>
>>> regex.match('1d 12h')
<re.Match object; span=(0, 2), match='1d'>
```

Reply:
I added unit tests to test all sort of cases of the function